[CI] Reorganize compile tests so new tests are automatically included in CI

gmagogsfm opened the pull request:
Previously unit tests in tests/compile are listed in CI config individually by file name, this means new test files won't automatically be included in CI. This PR creates a structure to categorize compile tests and updated CI config to run tests according to directory structure instead of individual files.

**New compile test folder structure**

- `compile/test_*.py` : various unit tests, future tests are most likely added here. New test files added here will be included in CI automatically
- `compile/fullgraph/` : full model tests, including all tests previously in `compile/piecewise`. New test files added here will be included in CI automatically
- `compile/distributed/` : tests that require multiple GPUs. New test files added here will **NOT** be included in CI automatically as these tests generally need to be manually configured to run in runners with particular number/type of GPUs.

Signed-off-by: Yanan Cao <gmagogsfm@gmail.com>

gmagogsfm committed:
Reorganize compile tests so new tests are automatically included in CI

Signed-off-by: Yanan Cao <gmagogsfm@gmail.com>

gmagogsfm committed:
Also update env_override.py and amd test

Signed-off-by: Yanan Cao <gmagogsfm@gmail.com>

zou3519 commented on .buildkite/test-amd.yaml:
```diff
@@ -187,7 +187,7 @@ steps:
   - tests/distributed/test_pynccl
   - tests/distributed/test_events
-  - tests/compile/test_basic_correctness
+  - tests/compile/fullgraph/test_basic_correctness.py
```
@gmagogsfm after this PR, t here are some tests in tests/compile/distributed, tests/compile/fullgraph, and also some individual tests under tests/compile (no subdirectory).

When would someone add a test to tests/compile/fullgraph vs tests/compile?

gmagogsfm replied:
I think `tests/compile` should mostly have unit tests, ie tests for a particular code path/feature

tests/compile/fullgraph are e2e full model tests that do not target specific features.

Their distinction mostly derives from previous setup where they had separate time budgets

```
- label: PyTorch Compilation Unit Tests # 15min
- label: PyTorch Fullgraph Smoke Test # 15min
- label: PyTorch Fullgraph Test # 27min
```

ProExpertProg reviewed (commented):
Thanks for this refactor, a few notes!

gmagogsfm committed:
Address comments

- Clean up old unused test file
- Add missing test dependencies
- Add tests/compile/README.md

Signed-off-by: Yanan Cao <gmagogsfm@gmail.com>

gmagogsfm commented on .buildkite/test-pipeline.yaml:
Done

gmagogsfm committed:
Fix broken relative import path after file moving

Signed-off-by: Yanan Cao <gmagogsfm@gmail.com>

zou3519 reviewed (approved):
sounds good, thank you for the readme

gmagogsfm committed:
Fix more relative import paths

Signed-off-by: Yanan Cao <gmagogsfm@gmail.com>

ProExpertProg commented:
Looks like a real failure this time about initializing CUDA in a forked subprocess

gmagogsfm committed:
Separate process for each test file

Signed-off-by: Yanan Cao <gmagogsfm@gmail.com>

ProExpertProg commented on .buildkite/test-pipeline.yaml:
```diff
@@ -466,9 +458,11 @@ steps:
+  # Run smoke tests under fullgraph directory, except test_full_graph.py
+  # as it is a heavy test that is covered in other steps.
+  # Use `find` to launch multiple instances of pytest so that
+  # they do not suffer from https://github.com/vllm-project/vllm/issues/21073
```
Can you add this comment to other places that use `find`?

gmagogsfm replied:
Added. I also filed a more appropriate issue for the problem with more context:
https://github.com/vllm-project/vllm/issues/28965

gmagogsfm committed:
Add comment clarifying use of find in test config

Signed-off-by: Yanan Cao <gmagogsfm@gmail.com>

ProExpertProg commented:
@gmagogsfm seems like there's a syntax issue in the YAML

gmagogsfm committed:
fix escaping character in yaml

Signed-off-by: Yanan Cao <gmagogsfm@gmail.com>

gmagogsfm committed:
add missing -

Signed-off-by: Yanan Cao <gmagogsfm@gmail.com>
